### PR TITLE
On Oculus Rift on/off, switch automatically to/from desktop mode

### PR DESCRIPTION
**Pull request by NeetBhagat**
Fix for Bug WL 21255 : On Oculus Rift on/off, switch automatically to/from desktop mode

 **Oculus Rift HMD / Handcontrollers required**

**Expected Result:**
When user put on Oculus Rift HMD, switch from desktop mode to VR mode.
And when the user removes the Oculus Rift HMD, switch from VR  to desktop mode.

_Note: Dev tested on Windows 10 + Oculus Rift._

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/173/HighFidelity-Beta-PR10548-a2851b5bcbe9db8147348e5d54543a666e2b690b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/173/HighFidelity-Beta-PR10548-a2851b5bcbe9db8147348e5d54543a666e2b690b.dmg

**Comment by OmegaHeron** [comment]
I'll QA this in approx 2 hours unless someone with Rift and controllers can get to it first.  I have required gear but have to finish another task first.

**Comment by OmegaHeron** [comment]
QA In progress.

**Comment by OmegaHeron** [comment]
Works as described - tested with Windows 10 - CV1 - Touch controllers.

QA Pass

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -1338,6 +1338,20 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
...
+			// then startOculusRiftStandBySession to listen Oculus HMD Mounted status. 
+			if (getActiveDisplayPlugin()->getName() != "Oculus Rift" && 
+				!oculusRiftPlugin->isStandBySessionActive()) {
+					startOculusRiftStandBySession();
+			}
+			// Poll periodically to check whether the user has worn Oculus HMD or not. And switch Display mode accordingly.
+			// If the user wear Oculus HMD then switch to VR mode. If the user removes Oculus HMD then switch to Desktop mode.
+			QTimer *switchDisplayModeTimer = new QTimer(this);
+			connect(switchDisplayModeTimer, SIGNAL(timeout()), this, SLOT(switchDisplayModeForOculus()));
+			switchDisplayModeTimer->start(500);
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: coding standard: no magic numbers, create a local named const.

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -6830,6 +6846,56 @@ void Application::updateDisplayMode() {
     Q_ASSERT_X(_displayPlugin, "Application::updateDisplayMode", "could not find an activated display plugin");
 }
 
+bool Application::isOculusRiftPluginAvailable() {
+	bool isOculusRiftPluginAvailable = false;
+	auto displayPlugins = PluginManager::getInstance()->getDisplayPlugins();
+	// Default to the first item on the list, in case none of the menu items match
+	DisplayPluginPointer defaultplugin = displayPlugins.at(0);
+	if (defaultplugin->isHmd() && defaultplugin->getName() == "Oculus Rift") {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: this name feels fragile, I'd suggest you use some common define in both the plugin and this code.

I also wonder if what we actually should do is introduce the idea that any display plugin could have this "waiting" feature... since surely some new HMD on the market will also support this idea.

File: libraries/plugins/src/plugins/Plugin.h [hunk 3]
```diff
@@ -73,6 +85,7 @@ class Plugin : public QObject {
 
 protected:
     bool _active { false };
+	bool _standbysessionstatus { false };
```
Comment (ZappoMan) [inline libraries/plugins/src/plugins/Plugin.h]: coding standard: indent

**Comment by ZappoMan** [comment]
To be clear, I really like this feature... I just think this implementation is a bit too fragile.

**Comment by tapalisa** [comment]
Sorry to come in with a design question this late- but isn't it likely that a VR user who removes the headset WANTS to be in pause mode in HMD, and not be switched to Desktop?

**Comment by OmegaHeron** [comment]
As an HMD user, oculus, I for one prefer it do switch back to desktop on removal, but, that said - options are a wonderful thing.  :)  If it was a config option of Go AFK on HMD remove or Go to Desktop then we could all likely be happy.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/234/HighFidelity-Beta-PR10548-0920dbc1a8910c986b2c83b2422aa7fa63e192d9.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/234/HighFidelity-Beta-PR10548-0920dbc1a8910c986b2c83b2422aa7fa63e192d9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/235/HighFidelity-Beta-PR10548-212627bd5dfb7ab27b0ef3c77ab2830b4017d4f0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/236/HighFidelity-Beta-PR10548-e90b1b5a67c8fd33d539c7d86b677ba24d17f61c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/236/HighFidelity-Beta-PR10548-e90b1b5a67c8fd33d539c7d86b677ba24d17f61c.dmg

**Comment by volansystech** [comment]
@ZappoMan , we made changes as you suggested. Please review and share your thoughts.

File: interface/src/Application.cpp [hunk 4]
```diff
@@ -1338,6 +1341,21 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
...
+            // Currently, autoswitch display mode support is only for Oculus Rift.
+            // If HMD Plugin is available and current display plugin is not HMD plugin
+            // then startHMDStandBySession to listen HMD Mounted status.
+            if (getActiveDisplayPlugin()->getName() != _hmdPluginName &&
+                !_hmdPlugin->isStandBySessionActive()) {
+                    startHMDStandBySession();
+            }
+            // Poll periodically to check whether the user has worn HMD or not. Switch Display mode accordingly.
+            // If the user wears HMD then switch to VR mode. If the user removes HMD then switch to Desktop mode.
+            QTimer *switchDisplayModeTimer = new QTimer(this);
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: coding standard: should be... `QTimer* switchDisplayModeTimer` with no space between type and `*` and a space between `*` and variable name.

File: interface/src/Application.cpp [hunk 5]
```diff
@@ -6732,6 +6752,50 @@ void Application::updateDisplayMode() {
     Q_ASSERT_X(_displayPlugin, "Application::updateDisplayMode", "could not find an activated display plugin");
 }
 
+bool Application::isHMDPluginAvailable() {
+    bool isHMDEnabledPluginAvailable = false;
+    auto displayPlugins = PluginManager::getInstance()->getDisplayPlugins();
+    // Currently, autoswitch display mode support is only for Oculus Rift.
+    foreach(DisplayPluginPointer displayPlugin, PluginManager::getInstance()->getDisplayPlugins()) {
+        if (displayPlugin->isHmd() && displayPlugin->getName() == OCULUS_RIFT_DISPLAY_PLUGIN_NAME) {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: this still seems to be Oculus dependent. It would be better for this API to allow other HMDs in the future that have this standby mode to also work.

Additionally - the name of this function is `isHMDPluginAvailable` which implies that it should return true for a Vive or a future Acer/MS HMD plugin... but it only rerturns true of an OCULUS rift plugin is available. This is confusing.

Additionally - you've introduced this new `_hmdPlugin` which is also misleading for the same reason.

What you *should* do is introduce the idea of some display plugins can determine if they are "being worn" - and if the display plugin detects that it's "being worn" then it does this auto-switch.

The problem with your implementation is:
- It's specific to Oculus... it should be generic for any plugin that can detect it's being worn.
- The variable names and function names refer to "HMD plugin" - but it really should be something about "being worn detecting" plugins (a mouthful, come up with a better name)...

File: interface/src/Application.h [hunk 6]
```diff
@@ -680,6 +680,14 @@ private slots:
     FileScriptingInterface* _fileDownload;
     AudioInjector* _snapshotSoundInjector { nullptr };
     SharedSoundPointer _snapshotSound;
+	
+	
+	DisplayPluginPointer _hmdPlugin; // HMD Enabled Plugin
```
Comment (ZappoMan) [inline interface/src/Application.h]: coding standard: indent should be 4 spaces

File: interface/src/Application.h [hunk 7]
```diff
@@ -680,6 +680,14 @@ private slots:
...
+    QString _hmdPluginName;
+    bool isHMDPluginAvailable();
+	bool _hmdMountedStatus; // Check HMD Mounted status
```
Comment (ZappoMan) [inline interface/src/Application.h]: coding standard: indent should be 4 spaces

File: interface/src/Application.h [hunk 8]
```diff
@@ -680,6 +680,14 @@ private slots:
...
+	bool startHMDStandBySession();
```
Comment (ZappoMan) [inline interface/src/Application.h]: coding standard: indent should be 4 spaces

File: interface/src/Application.h [hunk 9]
```diff
@@ -680,6 +680,14 @@ private slots:
...
+	void endHMDStandBySession();
```
Comment (ZappoMan) [inline interface/src/Application.h]: coding standard: indent should be 4 spaces

**Comment by NeetBhagat** [comment]
@ZappoMan , we have made changes as you suggested. Please review.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/261/HighFidelity-Beta-PR10548-58142817a12d132077515669953fd3c4087487df.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/261/HighFidelity-Beta-PR10548-58142817a12d132077515669953fd3c4087487df.dmg

File: interface/src/Application.cpp [hunk 10]
```diff
@@ -1338,6 +1341,40 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
         properties["active_display_plugin"] = getActiveDisplayPlugin()->getName();
         properties["using_hmd"] = isHMDMode();
 
+        _autoSwitchDisplayModeSupportedHMDPlugin = nullptr;
+        foreach(DisplayPluginPointer displayPlugin, PluginManager::getInstance()->getDisplayPlugins()) {
+            if (displayPlugin->isHmd() && 
+                displayPlugin->getName() == DISPLAY_PLUGIN_NAME_OCULUS_RIFT) {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: I really don't want you to do this approach... don't compare the plugin names.

Instead do something like:
- In DisplayPlugin.h add virtual method for `bool getSupportsAutoSwitch()` at the base return `false`.
- In the final implementation for the Oculus override that method and return `true`
- Change this code to be.... 
```
if (displayPlugin->isHmd() && displayPlugin->getSupportsAutoSwitch()) {
```

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/293/HighFidelity-Beta-PR10548-c7c07b28e94bc9d0b6aa0ac767e39aacb84c0b8c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/293/HighFidelity-Beta-PR10548-c7c07b28e94bc9d0b6aa0ac767e39aacb84c0b8c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/295/HighFidelity-Beta-PR10548-d27ad061cb003faaec06b66b856bec9b9c34b82c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/295/HighFidelity-Beta-PR10548-d27ad061cb003faaec06b66b856bec9b9c34b82c.dmg

**Comment by NeetBhagat** [comment]
Review comments of PR has been resolved. Please test and review.

File: interface/src/Application.cpp [hunk 11]
```diff
@@ -6732,6 +6766,32 @@ void Application::updateDisplayMode() {
     Q_ASSERT_X(_displayPlugin, "Application::updateDisplayMode", "could not find an activated display plugin");
 }
 
+void Application::switchDisplayMode() {
+    bool currentHMDWornStatus = _autoSwitchDisplayModeSupportedHMDPlugin->isDisplayVisible();
+    if (currentHMDWornStatus != _previousHMDWornStatus) {
+        // Switch to respective mode as soon as currentHMDWornStatus changes 
+        if (currentHMDWornStatus == false && _previousHMDWornStatus == true) {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: this should be ` if (!currentHMDWornStatus) {`

Since you can't be here unless they are opposite, no need to check the opposite value for `_previousHMDWornStatus`

File: interface/src/Application.cpp [hunk 12]
```diff
@@ -6732,6 +6766,32 @@ void Application::updateDisplayMode() {
...
+            qCDebug(interfaceapp) << "Switching from HMD to desktop mode";
+            setActiveDisplayPlugin(DESKTOP_DISPLAY_PLUGIN_NAME);
+            startHMDStandBySession();
+        }
+        if (currentHMDWornStatus == true && _previousHMDWornStatus == false) {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: this should be `} else {`

No need to check the `currentHMDWornStatus` or `_previousHMDWornStatus` because you can't be in this code if they aren't opposite, and you know that the earlier `if` means this must be `currentHMDWornStatus`

File: interface/src/Application.cpp (hunk 11 above)
Comment (ZappoMan) [inline interface/src/Application.cpp]: also - the coding standard says that you should have the nominal if case first... so really the if/else clauses should be flipped to be... `if (currentHMDWornStatus) { ... do the switch to being worn stuff... } else { ...do the switch to be taken off stuff... }`

File: interface/src/Application.cpp [hunk 13]
```diff
@@ -1338,6 +1340,35 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
...
+                    _autoSwitchDisplayModeSupportedHMDPlugin->getName();
+                _previousHMDWornStatus =
+                    _autoSwitchDisplayModeSupportedHMDPlugin->isDisplayVisible();
+                qCDebug(interfaceapp) << "_autoSwitchModeSupportedPluginName::" 
+                    << _autoSwitchDisplayModeSupportedHMDPluginName;
+                break;
+            }
+        }
+
+        if (_autoSwitchDisplayModeSupportedHMDPlugin != nullptr) {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: coding standard: should be `if (_autoSwitchDisplayModeSupportedHMDPlugin)`

File: interface/src/Application.cpp [hunk 14]
```diff
@@ -6732,6 +6766,32 @@ void Application::updateDisplayMode() {
     Q_ASSERT_X(_displayPlugin, "Application::updateDisplayMode", "could not find an activated display plugin");
 }
 
+void Application::switchDisplayMode() {
+    bool currentHMDWornStatus = _autoSwitchDisplayModeSupportedHMDPlugin->isDisplayVisible();
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: it looks to me like you don't guarantee that `_autoSwitchDisplayModeSupportedHMDPlugin` is a valid pointer. you should probably guard that here and exit early.

File: interface/src/Application.cpp [hunk 15]
```diff
@@ -1338,6 +1340,35 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
...
+                qCDebug(interfaceapp) << "_autoSwitchModeSupportedPluginName::" 
+                    << _autoSwitchDisplayModeSupportedHMDPluginName;
+                break;
+            }
+        }
+
+        if (_autoSwitchDisplayModeSupportedHMDPlugin != nullptr) {
+            // If HMD Plugin is available and current display plugin is not HMD plugin 
+            // then startHMDStandBySession to poll HMD Worn status.
+            if (getActiveDisplayPlugin()->getName() != _autoSwitchDisplayModeSupportedHMDPluginName &&
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: Do you really need to compare against the name of the plugin? Can't you just compare against the plugin pointer?

`if (getActiveDisplayPlugin() != _autoSwitchDisplayModeSupportedHMDPlugin ...`

File: interface/src/Application.cpp [hunk 16]
```diff
@@ -1567,7 +1598,10 @@ void Application::aboutToQuit() {
     }
 
     getActiveDisplayPlugin()->deactivate();
-
+    if (_autoSwitchDisplayModeSupportedHMDPlugin != nullptr
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: coding standard: should be.... `if (_autoSwitchDisplayModeSupportedHMDPlugin && _autoSwitchDisplayModeSupportedHMDPlugin->isStandBySessionActive()) {`

**Comment by tapalisa** [comment]
After trying this, I withdraw my design objection. Since the user is paused when a switch occurs, it's seamless and won't affect user's privacy (having their mic open). Sounds like there are some technical issues to resolve but seems like a good change otherwise!

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/350/HighFidelity-Beta-PR10548-e08c3453a44d912131ab6fa9aff62f1c1aa7837c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/350/HighFidelity-Beta-PR10548-e08c3453a44d912131ab6fa9aff62f1c1aa7837c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/354/HighFidelity-Beta-PR10548-6228ff19bf2c9315b6edc01747a985301421e436.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/354/HighFidelity-Beta-PR10548-6228ff19bf2c9315b6edc01747a985301421e436.dmg

**Comment by NeetBhagat** [comment]
Review comments of PR has been resolved. Please check.

**Comment by ZappoMan** [comment]
QA - please retest

**Comment by OmegaHeron** [comment]
Re-resting

**Comment by OmegaHeron** [comment]
QA Fail - there's a failure case here which may not have been considered.
HMD on
Open Tablet
Click Desktop (switch to desktop mode)
You remain in HMD mode.

This needs to honor tablet/hud switch to desktop mode selector.  Also this is not signalling down the chain, for lack of a better description, that mode has changed.  Once desktop mode is obtained via removing HMD the desktop HUD button 2nd from left should display HMD *not* Desktop.  HUD still thinks we're in HMD mode.

Basically every other menu/hud item can get out of sync with actual display in use.  For instance - take HMD off - click desktop HUD button for mode switch while HMD is *off head* and you get a mess.

**Comment by OmegaHeron** [comment]
Basically - to clear up my live notes as I was clicking buttons and formulating what's seen.

This automatic mode change works but *does not* remain synchronized with all the other ways one might toggle HMD/desktop -- this leads to some really screwed up in between cases once auto toggle has engaged and you subsequently use a tablet/hud/menu item to change mode.  It also does not honor mode change requests while HMD is on (a more cosmetic in feel problem, but, part of much greater issue of synch to display in use to *all* other elements that can change such).

**Comment by themelissabrown** [comment]
@ZappoMan can you comment on the design items above? Or should it go back to @PhilipRosedale for decisions?

**Comment by ZappoMan** [comment]
@themelissabrown - I think that the problem raised by @OmegaHeron are actually just bugs in the implementation. Basically, the implementation needs to make sure it handles mode changes from all possible vectors.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/439/HighFidelity-Beta-PR10548-ae9a928740719656f9a1bd40efef2b4a154335f2.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/439/HighFidelity-Beta-PR10548-ae9a928740719656f9a1bd40efef2b4a154335f2.dmg

**Comment by PhilipRosedale** [comment]
@themelissabrown @NeetBhagat Yes we need to make sure the state of the menu's and tablet matches the state that is automatically selected by this code.  Can you do that?

**Comment by NeetBhagat** [comment]
Resolved two issues:
1) While in HMD mode -> open Tablet -> Click Desktop.  Desktop mode will be on.
2) Menu sync problem while switching modes.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/447/HighFidelity-Beta-PR10548-c3621d7c325c634d2b53e09eec78641b4ed0454f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/447/HighFidelity-Beta-PR10548-c3621d7c325c634d2b53e09eec78641b4ed0454f.dmg

**Comment by OmegaHeron** [comment]
Windows 10 - CV1+Touch Controllers
QA Pass

HMD on/off toggles mode as expected.  HUD/Table indicators for mode switch work as expected and reflect correct current display state as HMD on/off automatic toggle invokes.  Also tested impact of removing HMD during first run tutorial - seems ok in that removing/replacing HMD did not break tutorial.
